<a href="https://colab.research.google.com/github/deltareticuli/colab/blob/main/hw2_petrov.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Petrov Ivan
## Variant 25

In [1]:
import sympy as sp
import numpy as np

from IPython.display import display, Math

## Task 1

Find the best approximation matrix $A_1$ of rank 2 of the matrix A in the norm $||\cdot||_2$ and find $||A-A_1||_2$, where  
$$A = \begin{bmatrix}
-6 & -64 & 0 & 64\\
-102 & -20 & 36 & -22\\
3 & 74 & -72 & 10
\end{bmatrix}$$

We need to find SVD decomposition of the matrix $A$:
$$ A = U\Sigma V^* $$

In [2]:
A = np.array([[-6, -64, 0, 64],
              [-102, -20, 36, -22],
              [3, 74, -72, 10]])

U, S, Vh = np.linalg.svd(A)

S_matrix = np.array([[S[0], 0, 0, 0],
                     [0, S[1], 0, 0],
                     [0, 0, S[2], 0]])

U, S_matrix, Vh

(array([[ 0.33333333, -0.66666667, -0.66666667],
        [ 0.66666667,  0.66666667, -0.33333333],
        [-0.66666667,  0.33333333, -0.66666667]]),
 array([[132.,   0.,   0.,   0.],
        [  0.,  99.,   0.,   0.],
        [  0.,   0.,  66.,   0.]]),
 array([[-5.45454545e-01, -6.36363636e-01,  5.45454545e-01,
          2.47237633e-16],
        [-6.36363636e-01,  5.45454545e-01,  5.15690493e-16,
         -5.45454545e-01],
        [ 5.45454545e-01, -8.58292749e-17,  5.45454545e-01,
         -6.36363636e-01],
        [ 8.45011042e-17,  5.45454545e-01,  6.36363636e-01,
          5.45454545e-01]]))

Let us check that the decomposition is correct:

In [3]:
sp.Matrix(np.matrix.round(U @ S_matrix @ Vh))

Matrix([
[  -6.0, -64.0,   0.0,  64.0],
[-102.0, -20.0,  36.0, -22.0],
[   3.0,  74.0, -72.0,  10.0]])

The rank of the matrix $A_1$ should be 2, so we need to construct a new matrix $\Sigma_1$ as follows: take the first two singular values and set the third one as zero:

In [4]:
S_1 = np.array([[S[0], 0, 0, 0],
                [0, S[1], 0, 0],
                [0, 0, 0, 0]])
sp.Matrix(S_1)

Matrix([
[132.0,  0.0, 0.0, 0.0],
[  0.0, 99.0, 0.0, 0.0],
[  0.0,  0.0, 0.0, 0.0]])

Now, let us compute matrix $A_1$:
$$ A_1 = U\Sigma_1 V^* $$

In [5]:
A_1 = U @ S_1 @ Vh
sp.Matrix(np.matrix.round(A_1))

Matrix([
[ 18.0, -64.0,  24.0,  36.0],
[-90.0, -20.0,  48.0, -36.0],
[ 27.0,  74.0, -48.0, -18.0]])

Finally, let us find $||A-A_1||_2$:

In [6]:
diff = A - np.matrix.round(A_1)
_, S_diff, _ = np.linalg.svd(diff)

result = max(S_diff)
result

66.0

## Answers:

$A_1 = \left[\begin{matrix}18 & -64 & 24 & 36\\-90 & -20 & 48 & -36\\27 & 74 & -48 & -18\end{matrix}\right]$


$||A-A_1||_2$ = 66

## Task 2

Estimate the relative error of the approximate solution $(1, 1)$ of the system $Ax = b$ in the norms $|\cdot|_1$ and $|\cdot|_2$ using the condition number of the matrix $A$, where
$$A = \begin{pmatrix} 4.91 & 0.15 \\8.18 & −8.11\end{pmatrix}, b = \begin{pmatrix} 4.95 \\−0.09\end{pmatrix}$$

$$\hat{A} = \begin{pmatrix} 5 & 0 \\8 & -8\end{pmatrix}, \hat{b} = \begin{pmatrix} 5 \\0\end{pmatrix}, \hat{x} = \begin{pmatrix} 1 \\1\end{pmatrix}$$

In [7]:
A = np.array([[5, 0],
              [8, -8]])
b = np.array([[5],
              [0]])
x = np.array([[1],
              [1]])

Let us find the inverse matrix of $\hat{A}$:

In [8]:
A_inv = np.linalg.inv(A)
sp.Matrix(A_inv)

Matrix([
[0.2,    0.0],
[0.2, -0.125]])

Now we need to find a condition number $\chi_1(A)$ for the norm $|\cdot|_1$:

$$\chi(A) \approx \chi(\hat{A}) \approx ||\hat{A}||\cdot||\hat{A}^{-1}||
$$

In [9]:
chi_1 = max(np.sum(np.abs(A), axis=0)) * max(np.sum(np.abs(A_inv), axis=0))
Math("\\chi_1(A) = %s" % chi_1)

<IPython.core.display.Math object>

Then let us find a condition number $\chi_2(A)$ for the norm $|\cdot|_2$:
$$\chi_2 = \sqrt{\frac{\lambda_{max}(A^*A)}{\lambda_{min}(A^*A)}}$$


Firstly, we need to find singular values for the matrix $A^*A$:

In [10]:
_, Sigma, _ = np.linalg.svd(A.T @ A)
print("Singular values: %s" % Sigma)

Singular values: [141.70927848  11.29072152]


In [11]:
chi_2 = np.sqrt(np.max(Sigma) / np.min(Sigma))
Math("\\chi_2(A) = %s \\approx %s" % (chi_2, np.round(chi_2, 2)))

<IPython.core.display.Math object>

Now, let us now find $\Delta b, \delta_1 b,\delta_2 b$:

In [12]:
b_1 = np.array([[4.95],
                [-0.09]])

Delta_b = b - b_1

delta_1_b = np.sum(np.abs(Delta_b)) / np.sum(np.abs(b))
delta_2_b = np.sqrt(np.sum(Delta_b**2)) / np.sqrt(np.sum(b**2))

In [13]:
Math("\\delta_1 b = %s" % np.round(delta_1_b, 3))

<IPython.core.display.Math object>

In [14]:
Math("\\delta_2 b = %s" % np.round(delta_2_b, 3))

<IPython.core.display.Math object>

And $\Delta A, \delta_1 A,\delta_2 A$:

In [15]:
A_1 = np.array([[4.91, 0.15],
                [8.18, -8.11]])

Delta_A = A - A_1

delta_1_A = np.max(np.sum(np.abs(Delta_A), axis=0)) / np.max(np.sum(np.abs(A), axis=0))
_, Sigma_delta, _ = np.linalg.svd(Delta_A.T @ Delta_A)
delta_2_A = np.sqrt(np.max(Sigma_delta) / np.max(Sigma))

In [16]:
Math("\\delta_1 A = %s" % np.round(delta_1_A, 3))

<IPython.core.display.Math object>

In [17]:
Math("\\delta_2 A = %s" % np.round(delta_2_A, 3))

<IPython.core.display.Math object>

In [18]:
Delta_b, np.sum(np.abs(Delta_b)), np.sum(np.abs(b)), np.sqrt(np.sum(Delta_b**2)), np.sqrt(np.sum(b**2))

(array([[0.05],
        [0.09]]), 0.13999999999999982, 5, 0.10295630140986992, 5.0)

In [19]:
Delta_A, np.max(np.sum(np.abs(Delta_A), axis=0)), np.max(np.sum(np.abs(A), axis=0)), np.max(Sigma_delta), np.max(Sigma)

(array([[ 0.09, -0.15],
        [-0.18,  0.11]]),
 0.2699999999999996,
 13,
 0.07098041280032272,
 141.70927848090332)

$$\Delta b = \begin{pmatrix}0.05 \\ 0.09\end{pmatrix}$$

$$\Delta A = \begin{pmatrix}0.09 & -0.15\\-0.18 & 0.11\end{pmatrix}$$

$$\delta_1 b =\frac{|\Delta b|_1}{|b|_1}\approx \frac{|\Delta b|_1}{|\hat{b}|_1} = \frac{0.14}{5}= 0.028$$

$$\delta_2 b =\frac{|\Delta b|_2}{|b|_2}\approx \frac{|\Delta b|_2}{|\hat{b}|_2} = \frac{0.1029}{5}= 0.021$$

$$\delta_1 A =\frac{|\Delta A|_1}{|A|_1}\approx \frac{|\Delta A|_1}{|\hat{A}|_1} = \frac{0.27}{13}= 0.021$$

$$\delta_2 A =\frac{|\Delta A|_2}{|A|_2}\approx \frac{|\Delta A|_2}{|\hat{A}|_2} =\frac{\sigma_1 (\Delta A)}{\sigma_1 (A)} = \frac{\sqrt{0.07098}}{\sqrt{141.70928}}= 0.022$$

In [20]:
delta_1_x_max = chi_1 * (delta_1_b + delta_1_A)
delta_2_x_max = chi_2 * (delta_2_b + delta_2_A)

delta_1_x_min = 1 / chi_1 * (delta_1_b + delta_1_A)
delta_2_x_min = 1 / chi_2 * (delta_2_b + delta_2_A)

Math("\delta_1 x \in [%s, %s]" % (np.round(delta_1_x_min, 3), np.round(delta_1_x_max, 3)))

<IPython.core.display.Math object>

In [21]:
Math("\delta_2 x \in [%s, %s]" % (np.round(delta_2_x_min, 3), np.round(delta_2_x_max, 3)))

<IPython.core.display.Math object>

In [22]:
delta_1_x_max - delta_1_x_min

0.24422130177514761

In [23]:
delta_2_x_max - delta_2_x_min

0.1401079299596348

## Answers:

$\delta_1 x \in [0.009, 0.254]$

$\delta_2 x \in [0.012, 0.152]$ - lowest relative error

## Task 3

Solve the system approximately and estimate the error of the solution in the norms $|\cdot|_1$, $|\cdot|_2$ and $|\cdot|_{\infty}$:
$$ 
\begin{cases} 
-(-8 + \varepsilon_1)x + 2(4 + \varepsilon_2)y = -5 + \varepsilon_3 \\ 
4x + (-3+\varepsilon_1)y = -2+ \varepsilon_4 
\end{cases}
$$
where the unknown numbers $\varepsilon_j$ satisfy the conditions $|\varepsilon_j| < 0.05$ for all $j=1,...,4$.

Firstly, let us write the given system in the matrix form:

$$A = \widehat A + \Delta A = \begin{pmatrix} 8 & 8 \\ 4 & -3 \end{pmatrix} + \begin{pmatrix} -\epsilon_{1} & 2\epsilon_{2} \\ 0 & \epsilon_{1} \end{pmatrix}$$

$$b = \widehat b + \Delta b = \begin{pmatrix} -5 \\ -2 \end{pmatrix} + \begin{pmatrix} \epsilon_{3} \\ \epsilon_{4} \end{pmatrix}$$

Then we need to find the inverse matrix of $\hat{A}$ and find the solution:

In [24]:
A = np.array([[8, 8],
              [4, -3]])
b = np.array([[-5], 
              [-2]])

A_inv = np.linalg.inv(A)
x = A_inv @ b

For the norm $|\cdot|_1$ the condition number is::

In [25]:
chi_1 = np.max(np.sum(np.abs(A), axis=0)) * np.max(np.sum(np.abs(A_inv), axis=0))
Math("\\chi_1(A) = %s \\approx %s" % (chi_1, np.round(chi_1, 2)))

<IPython.core.display.Math object>

For the norm $|\cdot|_\infty$ the condition number is:

In [26]:
chi_inf = np.max(np.sum(np.abs(A), axis=1)) * np.max(np.sum(np.abs(A_inv), axis=1))
Math("\\chi_\infty(A) = %s \\approx %s" % (chi_inf, np.round(chi_inf, 2)))

<IPython.core.display.Math object>

Let us find the condition number for the norm $|\cdot|_2$:

In [27]:
_, Sigma, _ = np.linalg.svd(A.T @ A)
print("Singular values: %s" % Sigma)

Singular values: [128.61765536  24.38234464]


In [28]:
chi_2 = np.sqrt(np.max(Sigma) / np.min(Sigma))
Math("\\chi_2(A) = %s \\approx %s" % (chi_2, np.round(chi_2, 2)))

<IPython.core.display.Math object>

In [29]:
e1, e2, e3, e4 = sp.symbols('varepsilon1:5')

A_1 = sp.Matrix([[8 - e1, 2 * (4 + e2)],
                 [4, -3 + e1]])
b_1 = sp.Matrix([[-5 + e3], 
                 [-2 + e4]])

Delta_A = A_1 - sp.Matrix(A)
Delta_b = b_1 - sp.Matrix(b)

Delta_A, Delta_b

(Matrix([
 [-varepsilon1, 2*varepsilon2],
 [           0,   varepsilon1]]), Matrix([
 [varepsilon3],
 [varepsilon4]]))

Let us find $\varepsilon_1, \varepsilon_2, \varepsilon_3, \varepsilon_4 \in \left[ -0.05, 0.05 \right]$ where $\delta_1 A + \delta_1 b$ is max.

In [30]:
epsilons = np.arange(-0.05, 0.05, 0.01)

result = []

for ep1 in epsilons:
    for ep2 in epsilons:
        for ep3 in epsilons:
            for ep4 in epsilons:
                Delta_b_iter = np.array(Delta_b.subs([(e3, ep3), (e4, ep4)])).astype(np.float64)
                Delta_A_iter = np.array(Delta_A.subs([(e1, ep1), (e2, ep2)])).astype(np.float64)
                
                delta_1_b = np.sum(np.abs(Delta_b_iter)) / np.sum(np.abs(b))
                delta_1_A = np.max(np.sum(np.abs(Delta_A_iter), axis=0)) / np.max(np.sum(np.abs(A), axis=0))
                delta_1_Ab = delta_1_A + delta_1_b
                
                result.append((delta_1_Ab, ep1, ep2, ep3, ep4))
                
print(max(result, key=lambda x: x[0]))

(0.026785714285714288, -0.05, -0.05, -0.05, -0.05)


So, $\varepsilon_1 = \varepsilon_2 = \varepsilon_3 = \varepsilon_4 = -0.05$

In [31]:
Delta_A = np.array(Delta_A.subs([(e1, -0.05), (e2, -0.05)])).astype(np.float64)
Delta_b = np.array(Delta_b.subs([(e3, -0.05), (e4, -0.05)])).astype(np.float64)

Delta_A, Delta_b

(array([[ 0.05, -0.1 ],
        [ 0.  , -0.05]]), array([[-0.05],
        [-0.05]]))

In the norm $|\cdot|_1$:

In [32]:
delta_1_b = np.sum(np.abs(Delta_b)) / np.sum(np.abs(b))
delta_1_A = np.max(np.sum(np.abs(Delta_A), axis=0)) / np.max(np.sum(np.abs(A), axis=0))

In [33]:
Math("\\delta_1 b = %s" % np.round(delta_1_b, 3))

<IPython.core.display.Math object>

In [34]:
Math("\\delta_1 A = %s" % np.round(delta_1_A, 3))

<IPython.core.display.Math object>

$$\delta_1 A =\frac{|\Delta A|_1}{|A|_1}\approx \frac{|\Delta A|_1}{|\hat{A}|_1} = 0.013$$
$$\delta_1 b =\frac{|\Delta b|_1}{|b|_1}\approx \frac{|\Delta b|_1}{|\hat{b}|_1} = 0.014$$

In the norm $|\cdot|_2$:

In [35]:
delta_2_b = np.sqrt(np.sum(Delta_b ** 2))/np.sqrt(np.sum(b**2))

_, Sigma_delta, _ = np.linalg.svd(Delta_A.T @ Delta_A)
delta_2_A = np.sqrt(np.max(Sigma_delta) / np.max(Sigma))

In [36]:
Math("\\delta_2 b = %s" % np.round(delta_2_b, 3))

<IPython.core.display.Math object>

In [37]:
Math("\\delta_2 A = %s" % np.round(delta_2_A, 3))

<IPython.core.display.Math object>

$$\delta_2 A =\frac{|\Delta A|_2}{|A|_2}\approx \frac{|\Delta A|_2}{|\hat{A}|_2} =\frac{\sigma_1 (\Delta A)}{\sigma_1 (A)} = 0.011$$
$$\delta_2 b =\frac{|\Delta b|_2}{|b|_2}\approx \frac{|\Delta b|_2}{|\hat{b}|_2} = 0.013$$

In the norm $|\cdot|_\infty$:

In [38]:
delta_inf_b = np.max(np.abs(Delta_b)) / np.max(np.abs(b))
delta_inf_A = np.max(np.sum(np.abs(Delta_A), axis=1)) / np.max(np.sum(np.abs(A), axis=1))

In [39]:
Math("\\delta_\infty b = %s" % np.round(delta_inf_b, 6))

<IPython.core.display.Math object>

In [40]:
Math("\\delta_\infty A = %s" % np.round(delta_inf_A, 6))

<IPython.core.display.Math object>

$$\delta_{\infty} A =\frac{|\Delta A|_{\infty}}{|A|_{\infty}}\approx \frac{|\Delta A|_{\infty}}{|\hat{A}|_{\infty}} = 0.009375$$
$$\delta_{\infty} b =\frac{|\Delta b|_{\infty}}{|b|_{\infty}}\approx \frac{|\Delta b|_{\infty}}{|\hat{b}|_{\infty}} = 0.01$$

In [41]:
delta_1_x_max = chi_1 * (delta_1_b + delta_1_A)
delta_2_x_max = chi_2 * (delta_2_b + delta_2_A)
delta_inf_x_max = chi_inf * (delta_inf_b + delta_inf_A)

np.round(delta_1_x_max, 6), np.round(delta_2_x_max, 6), np.round(delta_inf_x_max, 6)

(0.091837, 0.054604, 0.066429)

## Answers:

$\hat{x} = \left[\begin{matrix}-0.553571\\-0.071429\end{matrix}\right]$

$\delta_{1} x \leq 0.091837$

$\delta_{2} x \leq 0.054604$

$\delta_{\infty} x \leq 0.066429)$

## Task 4

Find the approximate inverse matrix to the matrix $A$ and evaluate the approximation error with respect to the uniform norm $||\cdot||_1$, if the elements of the matrix $A$ are known with an absolute error of 0.01:
$$A \approx \begin{pmatrix}-6 & -6\\-3 & 4\end{pmatrix}$$

Let us find the inverse matrix:

In [42]:
A = np.array([[-6, -6],
              [-3, 4]])

np.linalg.inv(A)

array([[-0.0952381 , -0.14285714],
       [-0.07142857,  0.14285714]])

Then we need to find the condition number in the uniform norm $||\cdot||_1$:

$$\chi_1(A) \approx \chi(\hat{A}) \approx ||\hat{A}||_1\cdot||\hat A^{-1}||_1=\max\{9,10\}\cdot \max\{0.167, 0.286\} = 10 \cdot 0.286 = 2.86
$$

$$||\epsilon||_{1} = ||\Delta A||_{1} = \displaystyle\max_{j}|\Delta A^{j}| = 0.02$$

$$\delta \varepsilon = \frac {||\varepsilon||_1}{||\hat A||_1}=\frac {0.02}{10}=0.002$$


Approximation error:
$$ \delta A^{-1} \leq \frac{\chi_1(\hat A) \delta \varepsilon}{1-\chi_1 (\hat A)\delta \varepsilon}=\frac{2.86\cdot0.002}{1-2.86 \cdot 0.002}=0.00575$$

## Answers:

$A^{-1} \approx \left[\begin{matrix}-0.09524 & -0.14286\\-0.07143 & 0.14286\end{matrix}\right]$

$ \delta A^{-1} \leq 0.00575$

## Task 5

Use simple iteration method for finding the solution of the given linear system 

$$\begin{cases} 22x + 8y + 1z = 6 \\ 5x + 23y + 5z = 5 \\ 5x + 1y + 20z = 1 \end{cases}$$

Determine the iteration number after which the approximation error for each coordinate does not exceed $0.01$ and find the corresponding approximate solution. Start with $x_0 = [0,0,0]^{T}$.

In [43]:
x_0 = np.array([[0],
                [0],
                [0]])
A = np.array([[22, 8, 1],
              [5, 23, 5],
              [5, 1, 20]])
B = np.array([[6],
              [5],
              [1]])

In [44]:
c_max = max(A[0][0], A[1][1], A[2][2])
C = np.array([[1 / c_max, 0, 0],
              [0, 1 / c_max, 0],
              [0, 0, 1 / c_max]])
C

array([[0.04347826, 0.        , 0.        ],
       [0.        , 0.04347826, 0.        ],
       [0.        , 0.        , 0.04347826]])

Let us find the matrix $P = I - CA$:

In [45]:
P = np.identity(3) - (C @ A)
P

array([[ 0.04347826, -0.34782609, -0.04347826],
       [-0.2173913 ,  0.        , -0.2173913 ],
       [-0.2173913 , -0.04347826,  0.13043478]])

Then we need to find $b=CB$:

In [46]:
b = C @ B 

$||P||_{\infty}=0.435<1$, it means that the iteration process converges.

In [47]:
np.max(np.sum(np.abs(P), axis=1))

0.43478260869565216

Now, let us find $x$ using the following formula: 
$$x^{k+1} = Px^{k} + b$$

In [48]:
x_1 = x_0
k = 0

while True:
    k += 1
    x_2 = P @ x_1 + b 
    
    if (np.abs(x_2 - x_1) < 0.01).all():
        break
    
    x_1 = x_2

sp.Matrix(np.round(x_1, 6))

Matrix([
[ 0.217473],
[ 0.178762],
[-0.007644]])

In [49]:
k

4

## Answers:

$x = \left[\begin{matrix}0.217473\\0.178762\\-0.007644\end{matrix}\right]$

After 4 iterations the approximation error for each coordinate does not exceed 0.01.

## Task 6

Find the most influential vertex in the graph:
$$A = \left[\begin{matrix}
0 & 1 & 0 & 1 & 0\\
1 & 1 & 1 & 0 & 0\\
1 & 1 & 1 & 0 & 0\\
0 & 1 & 1 & 1 & 0\\
0 & 0 & 0 & 1 & 1\end{matrix}\right]$$
using the PageRank algorithm with damping factor = $1 − β = 0.85$, where the graph adjacency matrix is defined as follows

In [50]:
A = np.array([[0, 1, 0, 1, 0],
              [1, 1, 1, 0, 0],
              [1, 1, 1, 0, 0],
              [0, 1, 1, 1, 0],
              [0, 0, 0, 1, 1]])

beta = 1 - 0.85

In [51]:
P = np.array([x / sum(row) for row in A for x in row]).reshape(5, 5)
sp.Matrix(P.round(3))

Matrix([
[  0.0,   0.5,   0.0,   0.5, 0.0],
[0.333, 0.333, 0.333,   0.0, 0.0],
[0.333, 0.333, 0.333,   0.0, 0.0],
[  0.0, 0.333, 0.333, 0.333, 0.0],
[  0.0,   0.0,   0.0,   0.5, 0.5]])

Now, let us build the matrix $Q$ and find the matrix $\hat{P}$:  

$$\hat P = P\cdot(1-\beta)+Q\cdot\beta,$$
where $\beta = 0.15$

In [52]:
Q = np.ones((5, 5)) / 5
P_hat = P * (1 - beta) + Q * beta

sp.Matrix(Q), sp.Matrix(P_hat)

(Matrix([
 [0.2, 0.2, 0.2, 0.2, 0.2],
 [0.2, 0.2, 0.2, 0.2, 0.2],
 [0.2, 0.2, 0.2, 0.2, 0.2],
 [0.2, 0.2, 0.2, 0.2, 0.2],
 [0.2, 0.2, 0.2, 0.2, 0.2]]), Matrix([
 [             0.03,             0.455,              0.03,             0.455,  0.03],
 [0.313333333333333, 0.313333333333333, 0.313333333333333,              0.03,  0.03],
 [0.313333333333333, 0.313333333333333, 0.313333333333333,              0.03,  0.03],
 [             0.03, 0.313333333333333, 0.313333333333333, 0.313333333333333,  0.03],
 [             0.03,              0.03,              0.03,             0.455, 0.455]]))

Initial state: $$X_0 = \left[\begin{matrix}0.2\\0.2\\0.2\\0.2\\0.2\end{matrix}\right]$$


In [53]:
x = np.ones(5) / 5
i = 1
while True:
    x_new = P_hat.T @ x
    display(Math("x_%s = [%s]^T" % (i, ", ".join(map(str, x_new.round(4))))))
    if (x_new - x < 0.001).all():
        break
    x = x_new
    i += 1

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

As it can be seen, the most influential vertex is vertex $2$.

## Answers:

## Task 7

Find the value $f(A)$ of the function $f(l) = e^{l+2}$, where 
$$A = \begin{pmatrix} 
-26 & 11 & -6 \\ 
-106 & 44 & -23 \\ 
-52 & 21 & -10 \end{pmatrix}$$

In [54]:
A = sp.Matrix([[-26, 11, -6],
               [-106, 44, -23],
               [-52, 21, -10]])

We need to find the Jordan form of the matrix A. In order to that, we need to find eigenvalues and eigenvectors first.

Let us find the determinant of the following matrix:

In [55]:
lambd = sp.symbols("lambda")
A - sp.eye(3) * lambd

Matrix([
[-lambda - 26,          11,           -6],
[        -106, 44 - lambda,          -23],
[         -52,          21, -lambda - 10]])

The determinant:

In [56]:
sp.det(A - sp.eye(3) * lambd)

-lambda**3 + 8*lambda**2 - 13*lambda + 6

Let us solve the following equation in order to get eigenvalues:

In [57]:
sp.Eq(sp.det(A - sp.eye(3) * lambd), 0)

Eq(-lambda**3 + 8*lambda**2 - 13*lambda + 6, 0)

In [58]:
eigenvalues = sp.solve(sp.det(A - sp.eye(3) * lambd))
eigenvalues

[1, 6]

In [59]:
x1, x2, x3 = sp.symbols('x1:4')
b = sp.Matrix([0, 0, 0])

(A - sp.eye(3) * eigenvalues[0], b)

(Matrix([
 [ -27, 11,  -6],
 [-106, 43, -23],
 [ -52, 21, -11]]), Matrix([
 [0],
 [0],
 [0]]))

First step, $\lambda = 1$, $det(A - I) = 0$:

In [60]:
eigenvector_1_1 = sp.linsolve((A - sp.eye(3) * eigenvalues[0], b), (x1, x2, x3))
eigenvector_1_1

FiniteSet((x3, 3*x3, x3))

Let us choose $x_3 = \frac{1}{5}$:

In [61]:
eigenvector_1_1 = eigenvector_1_1.subs(x3, sp.S(1) / 5)
eigenvector_1_1

FiniteSet((1/5, 3/5, 1/5))

Second step, $\lambda = 1$, $det(A - I)^2 = 0$:

In [62]:
eigenvector_1_2 = sp.linsolve(((A - sp.eye(3) * eigenvalues[0])**2, b), (x1, x2, x3))
eigenvector_1_2 

FiniteSet((2*x2/5 - x3/5, x2, x3))

Let us choose $x_2 = 1$ and $x_3 = 0$:

In [63]:
eigenvector_1_2 = eigenvector_1_2.subs(x2, 1).subs(x3, 0)
eigenvector_1_2

FiniteSet((2/5, 1, 0))

Third step, $\lambda = 6$, $det(A - 6I) = 0$:

In [64]:
eigenvector_2_1 = sp.linsolve((A - sp.eye(3) * eigenvalues[1], b), (x1, x2, x3))
eigenvector_2_1 

FiniteSet((x3/2, 2*x3, x3))

Let us choose $x_3 = 1$:

In [65]:
eigenvector_2_1 = eigenvector_2_1.subs(x3, 1)
eigenvector_2_1 

FiniteSet((1/2, 2, 1))

Finally, let us build the matrix T:

In [66]:
T = sp.Matrix([list(eigenvector_1_1)[0], list(eigenvector_1_2)[0], list(eigenvector_2_1)[0]]).T
T

Matrix([
[1/5, 2/5, 1/2],
[3/5,   1,   2],
[1/5,   0,   1]])

And the matrix T:

In [67]:
J = sp.Matrix([[1, 1, 0],
               [0, 1, 0],
               [0, 0, 6]])
J

Matrix([
[1, 1, 0],
[0, 1, 0],
[0, 0, 6]])

Let us check that the Jordan form is correct:

In [68]:
T*J*T**-1

Matrix([
[ -26, 11,  -6],
[-106, 44, -23],
[ -52, 21, -10]])

In [69]:
T*J*T**-1 == A

True

In order to find $f(A) = Tf(J)T^{-1}$ we need to find $f(J)$ first:

$f(J) = \left[\begin{matrix}f(1) & f'(1) & 0\\0 & f(1) & 0\\0 & 0 & f(6)\end{matrix}\right] = 
\left[\begin{matrix}e^3 & e^3 & 0\\0 & e^3 & 0\\0 & 0 & e^8\end{matrix}\right] $

Now we can find $f(A)$:

In [70]:
fJ = sp.Matrix([[sp.E**3, sp.E**3, 0],
               [0, sp.E**3, 0],
               [0, 0, sp.E**8]])

T*fJ*T**-1

Matrix([
[  -5*exp(8) + 4*exp(3),   -exp(3) + 2*exp(8),              -exp(8)],
[-20*exp(8) + 14*exp(3), -4*exp(3) + 8*exp(8),   -4*exp(8) + exp(3)],
[ -10*exp(8) + 8*exp(3), -3*exp(3) + 4*exp(8), -2*exp(8) + 2*exp(3)]])

$f(A) = \left[\begin{matrix}\frac{1}{5} & \frac{2}{5} & \frac{1}{2}\\\frac{3}{5} & 1 & 2\\\frac{1}{5} & 0 & 1\end{matrix}\right]\left[\begin{matrix}e^3 & e^3 & 0\\0 & e^3 & 0\\0 & 0 & e^8\end{matrix}\right]\left[\begin{matrix}50 & -20 & 15\\-10 & 5 & -5\\-10 & 4 & -2\end{matrix}\right] = \left[\begin{matrix}-14824.4477875159 & 5941.83043716027 & -2980.95798704173\\-59337.9622239099 & 23767.3217486411 & -11903.7464112437\\-29648.8955750318 & 11863.5753373973 & -5921.74490023708\end{matrix}\right]
$

## Answers:

$f(A) = \left[\begin{matrix}-14824.4477875159 & 5941.83043716027 & -2980.95798704173\\-59337.9622239099 & 23767.3217486411 & -11903.7464112437\\-29648.8955750318 & 11863.5753373973 & -5921.74490023708\end{matrix}\right]$